src/libretro_ui.cpp

```diff
@@ -0,0 +1,739 @@
+	info->need_fullpath = true;
+	info->valid_extensions = "ldb";
+}
+
+/* Gets information about system audio/video timings and geometry.
+ * Can be called only after retro_load_game() has successfully completed.
+ * NOTE: The implementation of this function might not initialize every
+ * variable if needed.
+ * E.g. geom.aspect_ratio might not be initialized if core doesn't
+ * desire a particular aspect ratio. */
+RETRO_API void retro_get_system_av_info(struct retro_system_av_info* info) {
+	info->geometry.base_width = SCREEN_TARGET_WIDTH;
+	info->geometry.base_height = SCREEN_TARGET_HEIGHT;
+	info->geometry.max_width = SCREEN_TARGET_WIDTH;
+	info->geometry.max_height = SCREEN_TARGET_HEIGHT;
+	info->geometry.aspect_ratio = 0.0f;
+	info->timing.fps = Graphics::GetDefaultFps();
+	info->timing.sample_rate = AUDIO_SAMPLERATE;
+}
+
+/* Sets device to be used for player 'port'.
+ * By default, RETRO_DEVICE_JOYPAD is assumed to be plugged into all
+ * available ports.
+ * Setting a particular device type is not a guarantee that libretro cores
+ * will only poll input based on that particular device type. It is only a
+ * hint to the libretro core when a core cannot automatically detect the
+ * appropriate input device type on its own. It is also relevant when a
+ * core can change its behavior depending on device type. */
+RETRO_API void retro_set_controller_port_device(unsigned port, unsigned device) {
+	// Not used
+}
+
+/* Resets the current game. */
+RETRO_API void retro_reset(void) {
+	Player::reset_flag = true;
+}
+
+/* Runs the game for one video frame.
+ * During retro_run(), input_poll callback must be called at least once.
+ *
+ * If a frame is not rendered for reasons where a game "dropped" a frame,
+ * this still counts as a frame, and retro_run() should explicitly dupe
+ * a frame if GET_CAN_DUPE returns true.
+ * In this case, the video callback can take a NULL argument for data.
+ */
+
+RETRO_API void retro_run() {
+	Player::MainLoop();
+
+	if (!DisplayUi) {
+		// Player::Exit was called, send shutdown request to the frontend
+		LibretroUi::player_exit_called = true;
+		// This closes the whole frontend and not just the core, but there is no API for core unloading...
+		LibretroUi::environ_cb(RETRO_ENVIRONMENT_SHUTDOWN, nullptr);
+	}
+}
+
+static void extract_directory(char *buf, const char *path, size_t size) {
+	strncpy(buf, path, size - 1);
+	buf[size - 1] = '\0';
+
+	char *base = strrchr(buf, '/');
+	if (!base)
+		base = strrchr(buf, '\\');
+
+	if (base)
+		*base = '\0';
+	else
+		buf[0] = '\0';
+}
+
+/* Loads a game. */
+RETRO_API bool retro_load_game(const struct retro_game_info* game) {
+	char parent_dir[1024];
+	enum retro_pixel_format fmt = RETRO_PIXEL_FORMAT_XRGB8888;
+
+	if (!game)
+		return false;
+
+	if (!LibretroUi::environ_cb(RETRO_ENVIRONMENT_SET_PIXEL_FORMAT, &fmt)) {
+		log_cb(RETRO_LOG_INFO, "XRGB8888 is not supported.\n");
+		return false;
+	}
+
+	extract_directory(parent_dir, game->path, sizeof(parent_dir));
```

Comment:
This crashes on Windows because path in game->path is not UTF-8. Don't tell me they use the local Codepage here m(

Reply:
Nevermind, the path is UTF-8 encoded but liblcf can't handle such paths on Windows. In the GameBrowser we work around this with "SetCurrentDirectory" but I have a proper solution for this in my VFS branch, so I won't fix this. This is only a problem for the filename of the main game folder (where liblcf must have access to).